sqlglot/dialects.py

```diff
@@ -113,18 +113,18 @@ def _unix_to_time(self, expression):
     }
 
     functions = {
-        'APPROX_COUNT_DISTINCT': lambda args: exp.ApproxDistinct(this=args[0]),
-        'EPOCH': lambda args: exp.TimeToUnix(this=args[0]),
+        'APPROX_COUNT_DISTINCT': exp.ApproxDistinct.from_arg_list,
+        'EPOCH': exp.TimeToUnix.from_arg_list,
         'EPOCH_MS': lambda args: exp.UnixToTime(this=exp.Slash(
-            this=args[0],
+            this=list_get(args, 0),
```

Comment:
@tobymao here and below I switch from accessing an argument by index directly to using the `list_get` help to avoid the generic `IndexError: list index out of range` error which leads to a loss of the parser debugging info.

Reply:
Now I'm thinking it was the wrong thing to do provided the existing validation only checks whether the key was set and not whether it was None or not ([code](https://github.com/izeigerman/sqlglot/blob/main/sqlglot/expressions.py#L79))

Reply:
it should probably be fine?

Reply:
Not really. The semantics has changed. Previously we failed with `IndexError`. Now we don't, since the `list_get` returns `None` if the argument is missing. However the existing validation only catches missing arguments if the key is missing (and ignores the scenario where it's set to `None` explicitly). With this change we always explicitly set the key even if the argument is missing in the list.

Reply:
Additionally I'm overall confused by how missing arguments are handled. For example in for the [Join](https://github.com/tobymao/sqlglot/blob/main/sqlglot/expressions.py#L234) expression we expect the `on` argument to be mandatory. However we violate this assumption when we handle `CROSS JOIN`. We work this around by actually providing the `on` key, but setting it to `None`.

Reply:
i think missing arguments just means that it's required to set it. i think you may be able to change this to check that the value is not none? for example

args.get(key) is not None

Reply:
I've tried that and a bunch of tests failed :) 
```
======================================================================
ERROR: test_spark (tests.test_dialects.TestDialects)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/izeigerman/github/sqlglot/tests/test_dialects.py", line 517, in test_spark
    transpile(
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 24, in transpile
    for expression in parse(code, read, error_level=error_level)
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 13, in parse
    return dialect.parse(code, **opts)
  File "/Users/izeigerman/github/sqlglot/sqlglot/dialects.py", line 17, in parse
    return self.parser(**opts).parse(self.tokenizer().tokenize(code), code)
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 193, in parse
    expressions.append(self._parse_statement())
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 261, in _parse_statement
    cte = self._parse_cte()
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 416, in _parse_cte
    return exp.CTE(
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 14, in __init__
    self.validate()
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 81, in validate
    raise ValueError(f"Required keyword: {k} missing for {self.token_type}")
ValueError: Required keyword: this missing for TokenType.WITH

======================================================================
ERROR: test_ctas (tests.test_rewriter.TestRewriter)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/izeigerman/github/sqlglot/tests/test_rewriter.py", line 11, in test_ctas
    Rewriter(expression).ctas('x').expression.sql(),
  File "/Users/izeigerman/github/sqlglot/sqlglot/rewriter.py", line 13, in wrapper
    expression = self.func(rewriter, *args, **kwargs)
  File "/Users/izeigerman/github/sqlglot/sqlglot/rewriter.py", line 37, in ctas
    file_format = exp.FileFormat(this=file_format),
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 14, in __init__
    self.validate()
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 81, in validate
    raise ValueError(f"Required keyword: {k} missing for {self.token_type}")
ValueError: Required keyword: this missing for TokenType.FORMAT

======================================================================
ERROR: test_identity (tests.test_transpile.TestTranspile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/izeigerman/github/sqlglot/tests/test_transpile.py", line 181, in test_identity
    self.assertEqual(transpile(sql)[0], sql.strip())
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 24, in transpile
    for expression in parse(code, read, error_level=error_level)
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 13, in parse
    return dialect.parse(code, **opts)
  File "/Users/izeigerman/github/sqlglot/sqlglot/dialects.py", line 17, in parse
    return self.parser(**opts).parse(self.tokenizer().tokenize(code), code)
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 193, in parse
    expressions.append(self._parse_statement())
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 261, in _parse_statement
    cte = self._parse_cte()
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 397, in _parse_cte
    return self._parse_select()
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 429, in _parse_select
    'from': self._parse_from(),
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 456, in _parse_from
    return exp.From(expressions=self._parse_csv(self._parse_table))
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 941, in _parse_csv
    items = [parse()]
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 530, in _parse_table
    this = exp.Alias(this=this, alias=None)
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 14, in __init__
    self.validate()
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 81, in validate
    raise ValueError(f"Required keyword: {k} missing for {self.token_type}")
ValueError: Required keyword: alias missing for TokenType.ALIAS

======================================================================
ERROR: test_partial (tests.test_transpile.TestTranspile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/izeigerman/github/sqlglot/tests/test_transpile.py", line 186, in test_partial
    self.assertEqual(transpile(sql, error_level=ErrorLevel.IGNORE)[0], sql.strip())
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 24, in transpile
    for expression in parse(code, read, error_level=error_level)
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 13, in parse
    return dialect.parse(code, **opts)
  File "/Users/izeigerman/github/sqlglot/sqlglot/dialects.py", line 17, in parse
    return self.parser(**opts).parse(self.tokenizer().tokenize(code), code)
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 193, in parse
    expressions.append(self._parse_statement())
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 261, in _parse_statement
    cte = self._parse_cte()
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 397, in _parse_cte
    return self._parse_select()
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 429, in _parse_select
    'from': self._parse_from(),
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 456, in _parse_from
    return exp.From(expressions=self._parse_csv(self._parse_table))
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 941, in _parse_csv
    items = [parse()]
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 520, in _parse_table
    expression = exp.Table(this=table, db=db)
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 14, in __init__
    self.validate()
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 81, in validate
    raise ValueError(f"Required keyword: {k} missing for {self.token_type}")
ValueError: Required keyword: this missing for TokenType.TABLE

----------------------------------------------------------------------
Ran 31 tests in 0.213s

FAILED (errors=4)
```

Reply:
Perhaps I should do this validation for functions only

Reply:
can we fix those expressions to not have those args required?

Reply:
I can try that! Just wasn't sure whether this will break some other assumption. It's hard to assess the impact.

Reply:
@tobymao I've refactored the error handling a bit and made it more consistent. Now the following values are treated as missing arguments:
```
None
[]
[None]  # Captured a couple of those as well
```
Please check out the corresponding [commit](https://github.com/tobymao/sqlglot/pull/25/commits/127660b6c94946f20aa6261e05ec13e78474e1a9). Please note that with this change the parsing of badly formed expressions no longer succeeds like it did before. Instead the exception is thrown or no expression is returned if the `ErrorLevel` is set to `IGNORE`.